File: packages/react-native-v2/src/specs/NativeEvervault.ts
import { TurboModule, TurboModuleRegistry } from "react-native";

export interface Spec extends TurboModule {
  /**
   * Initialize the Evervault SDK.
   *
   * @param teamId - The team ID.
   * @param appId - The app ID.
   *
   * @returns A unique identifier for the SDK instance.
   */
  initialize(teamId: string, appId: string): string;

  /**
   * Encrypt a string.
   *
   * @param instanceId - The unique identifier for the SDK instance.
   * @param data - The string to encrypt.
   *
   * @returns The encrypted string.
   */
  encryptString(instanceId: string, data: string): Promise<string>;

  /**
   * Encrypt a number.
   *
   * @param instanceId - The unique identifier for the SDK instance.
   * @param data - The number to encrypt.
   *
   * @returns The encrypted number.
   */
  encryptNumber(instanceId: string, data: number): Promise<string>;

  /**
   * Encrypt a boolean.
   *
   * @param instanceId - The unique identifier for the SDK instance.
   * @param data - The boolean to encrypt.
   *
   * @returns The encrypted boolean.
   */
  encryptBoolean(instanceId: string, data: boolean): Promise<string>;

  /**
   * Encrypt an object.
   *
   * @param instanceId - The unique identifier for the SDK instance.
   * @param data - The object to encrypt.
   *
   * @returns The encrypted object.
   */
  encryptObject(instanceId: string, data: Object): Promise<Object>;

  /**
   * Encrypt an array.
   *
   * @param instanceId - The unique identifier for the SDK instance.
   * @param data - The array to encrypt.
   *
   * @returns The encrypted array.
   */
  encryptArray(instanceId: string, data: Array<any>): Promise<Array<any>>;
}

export const NativeEvervault = TurboModuleRegistry.get<Spec>(
  "NativeEvervault"
) as Spec | null;

File: packages/react-native-v2/ios/NativeEvervault.swift
import Evervault

@objc(NativeEvervault)
class NativeEvervault: NSObject {
    var instances: [String: Evervault] = [:]

    func getInstanceId(_ teamId: String, _ appId: String) -> String {
        return "\(teamId):\(appId)"
    }

    func getInstance(_ instanceId: String) -> Evervault? {
        return self.instances[instanceId]
    }

    @objc(initialize:withAppId:)
    func initialize(teamId: String, appId: String) -> String {
        let evervault = Evervault(teamId: teamId, appId: appId)
        let instanceId = getInstanceId(teamId, appId)
        self.instances[instanceId] = evervault
        return instanceId
    }

    func encrypt(instanceId: String, _ value: Any, _ resolve: @escaping RCTPromiseResolveBlock, _ reject: @escaping RCTPromiseRejectBlock) {
        Task {
            do {
                guard let instance = getInstance(instanceId) else {
                    reject("Error", "Instance not found", nil)
                    return
                }
                let enc = try await instance.encrypt(value)
                resolve(enc)
            } catch {
                reject("Error", "Failed to encrypt: \(error.localizedDescription)", error)
            }
        }
    }

    @objc(encryptString:withValue:withResolver:withRejecter:)
    func encryptString(instanceId: String, value: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        return self.encrypt(instanceId: instanceId, value, resolve, reject)
    }

    @objc(encryptNumber:withValue:withResolver:withRejecter:)
    func encryptNumber(instanceId: String, value: Double, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        return self.encrypt(instanceId: instanceId, value, resolve, reject)
    }

    @objc(encryptBoolean:withValue:withResolver:withRejecter:)
    func encryptBoolean(instanceId: String, value: Bool, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        return self.encrypt(instanceId: instanceId, value, resolve, reject)
    }

    @objc(encryptObject:withValue:withResolver:withRejecter:)
    func encryptObject(instanceId: String, value: [String: Any], resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        return self.encrypt(instanceId: instanceId, value, resolve, reject)
    }

    @objc(encryptArray:withValue:withResolver:withRejecter:)
    func encryptArray(instanceId: String, value: [Any], resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        return self.encrypt(instanceId: instanceId, value, resolve, reject)
    }
}
